lib/nypl_sierra_api_client.rb

```diff
@@ -55,49 +57,66 @@ def do_request (method, path, options = {})
 
     authenticate! if options[:authenticated]
 
-    uri = URI.parse("#{@config[:base_url]}#{path}")
-
-    http = Net::HTTP.new(uri.host, uri.port)
-    http.use_ssl = uri.scheme === 'https'
+    @uri = URI.parse("#{@config[:base_url]}#{path}")
 
     # Build request headers:
     request_headers = {}
     request_headers['Content-Type'] = options[:headers]['Content-Type'] unless options.dig(:headers, 'Content-Type').nil?
 
     # Create HTTP::Get or HTTP::Post
-    request =  Net::HTTP.const_get(method.capitalize).new(uri, request_headers)
+    request =  Net::HTTP.const_get(method.capitalize).new(@uri, request_headers)
 
     # Add bearer token header
     request['Authorization'] = "Bearer #{@access_token}" if options[:authenticated]
 
     # Allow caller to modify the request before we send it off:
     yield request if block_given?
 
-    logger.debug "SierraApiClient: #{method} to Sierra api", { uri: uri, body: request.body }
+    logger.debug "SierraApiClient: #{method} to Sierra api", { uri: @uri, body: request.body }
+
+    execute request
+  end
+
+  def execute (request)
+    http = Net::HTTP.new(@uri.host, @uri.port)
+    http.use_ssl = @uri.scheme === 'https'
 
     begin
-      # Execute request:
       response = http.request(request)
+      logger.debug "SierraApiClient: Got Sierra api response", { code: response.code, body: response.body }
     rescue => e
-      raise SierraApiClientError.new(e), "Failed to #{method} to #{path}: #{e.message}"
+      raise SierraApiClientError.new "Failed to #{request.method} to #{request.path}: #{e.message}"
     end
 
-    logger.debug "SierraApiClient: Got Sierra api response", { code: response.code, body: response.body }
-
-    parse_response response
+    handle_response response, request
   end
 
-  def parse_response (response)
+  def handle_response (response, request)
     if response.code == "401"
       # Likely an expired access-token; Wipe it for next run
-      # TODO: Implement token refresh
+
       @access_token = nil
-      raise SierraApiClientTokenError.new("Got a 401: #{response.body}")
+      logger.debug "SierraApiClient: Unathorized", { code: 401, body: response.body }
+      reattempt request
     end
 
+    reset_retries if @retries > 0
     SierraApiResponse.new(response)
   end
 
+  def reattempt request
+    raise SierraApiClientTokenError.new("Maximum retries exceeded") if @retries >= 3
```

Comment:
I think when talking about this earlier I said the retries count should be zeroed on success, but I'm now realizing it should also be reset on max failure. So, in addition to raising this, I think you may want to call `reset_retries` so that the slate is clean for the next call. Otherwise, a call that maxes out retries will cause the next call to effectively get only one retry.